packages/integration/src/azure/CachedAzureDevOpsCredentialsProvider.ts

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2023 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { AzureDevOpsCredential, PersonalAccessTokenCredential } from './config';
+import {
+  ClientSecretCredential,
+  ManagedIdentityCredential,
+  TokenCredential,
+} from '@azure/identity';
+import {
+  AzureDevOpsCredentials,
+  AzureDevOpsCredentialsProvider,
+} from './types';
+
+type CachedAzureDevOpsCredentials = AzureDevOpsCredentials & {
+  expiresAt?: number;
+};
+
+const tenMinutes = 1000 * 60 * 10;
+
+/**
+ * A credentials provider that caches the credentials for as long as it is valid.
+ *
+ * @public
+ */
+export class CachedAzureDevOpsCredentialsProvider
+  implements AzureDevOpsCredentialsProvider
+{
+  azureDevOpsScope = '499b84ac-1321-427f-aa17-267ca6975798/.default';
+  cached: CachedAzureDevOpsCredentials | undefined;
+
+  static fromAzureDevOpsCredential(
+    credential: AzureDevOpsCredential,
+  ): CachedAzureDevOpsCredentialsProvider {
+    switch (credential.kind) {
+      case 'PersonalAccessToken':
+        return CachedAzureDevOpsCredentialsProvider.fromPersonalAccessTokenCredential(
+          credential,
+        );
+      case 'ClientSecret':
+        return CachedAzureDevOpsCredentialsProvider.fromTokenCredential(
+          new ClientSecretCredential(
+            credential.tenantId,
+            credential.clientId,
+            credential.clientSecret,
+          ),
+        );
+      case 'ManagedIdentity':
+        return CachedAzureDevOpsCredentialsProvider.fromTokenCredential(
+          new ManagedIdentityCredential(credential.clientId),
+        );
+      default:
+        // @ts-ignore
+        // eslint-disable-next-line no-case-declarations
+        const _exhaustiveCheck: never = credential;
```

Comment:
Could you refactor this to note require a `@ts-ignore`? For example just making it an if-else chain, or I think `switch (credential.kind as string) {` might do it too.

Reply:
The `@ts-ignore` was there because `_exhaustiveCheck` was declared but never used. And  `_exhaustiveCheck: never` was there to generate a typescript error when you forget to cover a (new) case in the switch statement.

I was able to get rid of the `@ts-ignore` and as a bonus the `eslint-disable-next-line` by declaring a function for the exhaustive type check instead:
```diff
+ function exhaustiveCheck(_param: never) {}
....
- // @ts-ignore
- // eslint-disable-next-line no-case-declarations
- const _exhaustiveCheck: never = credential;
+ exhaustiveCheck(credential);
```